### PR TITLE
Strict focusing using Default Goal Selector.

### LINKED ISSUE #6689: Feature request: even stricter subproofs
`Set Bullet Behavior` should accept an argument named something like `"Enforce Subproofs"`. When this option is active, proofs must be bullet-structured throughout. That is, whenever more than one goal is in focus, a new level of bullet must precede the use of any tactics.

### PR DESCRIPTION
**Pull request by SkySkimmer**
We add a [SelectAlreadyFocused] constructor to [goal_selector] (read
"!") which causes a failure when there's not exactly 1 goal and
otherwise is a noop.

Then [Set Default Goal Selector "!".] puts us in "strict focusing"
mode where we can only run tactics if we have only one goal or use a
selector.

Closes #6689.

### CODE REVIEW

File: proofs/pfedit.ml
Comment (Zimmi48) [inline proofs/pfedit.ml]: As you said it yourself, this error message should be improved.

File: test-suite/success/goal_selector.v
Comment (Zimmi48) [inline test-suite/success/goal_selector.v]: You should also test `only !:` and it would probably be good that some of these tests are done before the `Set Default Goal Selector "!"`

File: tactics/tacticals.ml
Comment (Zimmi48) [inline tactics/tacticals.ml]: Same error message: should also be improved.

**Review by Zimmi48 — APPROVED** [review]
I have no particular opinion about the syntax. I imagine the most common use should be in setting the default goal selector. Thus, it is not really necessary to have such a short syntax.

Apart from the things that you listed yourself as TODOs (documentation, CHANGES, better error message) and a small remark on the test case, this PR looks very good to me.

File: test-suite/success/goal_selector.v
Comment (SkySkimmer) [reply test-suite/success/goal_selector.v]: `only !:` doesn't parse (same as `only all:`). I think it would be the same as `;[tac]` so there's no need to add it.

Comment (Zimmi48) [reply test-suite/success/goal_selector.v]: @cmangin is the specialist here and should confirm this but then I think that your update to `tclSELECT` is not necessary: you could just as well raise an anomaly.

Comment (cmangin) [reply test-suite/success/goal_selector.v]: I think @Zimmi48 is right, it should even be an anomaly for the `SelectAll` case, probably.
(I guess ideally, the notion of toplevel selector and `only` selector should be distinguished, just as it is in the parsing, but maybe it's not worth the trouble.)

**Comment by SkySkimmer** [comment]
Updated.

**Review by aspiwack — APPROVED** [review]
The code is very simple. It's all good to me.

I really like the idea. Thanks!

**Comment by SkySkimmer** [comment]
>I really like the idea.

https://github.com/coq/coq/issues/6689#issuecomment-362932955
>This is an old idea that was never implemented. @aspiwack proposed to implement this by having a new goal selector e.g. unique: which would apply a tactic to the current goal if it's the only one under focus and fail otherwise. Setting this goal selector as the default goal selector would then be enough to obtain the requested behavior.

**Comment by aspiwack** [comment]
I'm aware :-) . Don't spoil a compliment!

**Review by Zimmi48 — APPROVED** [review]
I re-approve the current form fully :)

**Comment by Zimmi48** [comment]
@ppedrot You're the only relevant code owner here.

File: proofs/pfedit.ml
Comment (ppedrot) [inline proofs/pfedit.ml]: Even though observationally equivalent in this context, I'd rather use `tclZEROMSG` rather than raising an exception here, it'd be safer if ever the surrounding code gets modified.

Comment (SkySkimmer) [reply proofs/pfedit.ml]: >Even though observationally equivalent in this context, I'd rather use tclZEROMSG rather than raising an exception here, it'd be safer if ever the surrounding code gets modified.

OK but can you explain more about how it's safer?

Comment (ppedrot) [reply proofs/pfedit.ml]: This currently works because the only thing you do with the resulting monadic value is to run it. But if you were to compose it with other code, due to the very higher-orderedness of the monad, the exception would get raised at an unpredictible time and would most likely wreak havoc w.r.t. intended semantics. Except for very specific cases that are documented (e.g. `Goal.enter`) one should not raise exceptions inside the monadic layer.

Comment (SkySkimmer) [reply proofs/pfedit.ml]: ZEROMSG is not available in pfedit x_x

Comment (ppedrot) [reply proofs/pfedit.ml]: Well, you can always use `tclZERO` + `UserError`...

Comment (SkySkimmer) [reply proofs/pfedit.ml]: >Well, you can always use tclZERO + UserError...

That's what I did.

**Comment by Zimmi48** [comment]
The SF failure is unrelated. If you want an all green CI you can rebase on master.

File: proofs/pfedit.ml
Comment (ppedrot) [reply proofs/pfedit.ml]: No, you used `user_error` which is just a wrapper raising an exception. You should use a monad combinator instead.

Comment (SkySkimmer) [reply proofs/pfedit.ml]: The PR has been updated since then.

**Review by ppedrot — APPROVED** [review]
Good to go.

**Comment by ppedrot** [comment]
Do we agree I don't have to wait for reviews from other owners to merge this PR, as they're not relevant? *Chat échaudé craint l'eau froide.*

**Comment by maximedenes** [comment]
I want to review the documentation, but it'll have to wait until tomorrow, as today is not a working day here.

**Comment by ppedrot** [comment]
@maximedenes Technically it's not here either, but there is no better day than holidays to hack Coq!

**Comment by ppedrot** [comment]
@maximedenes Waiting on your documentation review.

**Comment by SkySkimmer** [comment]
Rebased (there was a doc conflict)

**Review by maximedenes — APPROVED** [review]
The documentation looks ok to me.

**Comment by ppedrot** [comment]
@SkySkimmer Can you rebase to that I can merge this PR? Thanks.

**Comment by SkySkimmer** [comment]
Weirdly I encountered no conflict when rebasing. I also tried a merge and still no conflict. Github is making them up I guess.
Rebased regardless.

**Comment by ejgallego** [comment]
> Weirdly I encountered no conflict when rebasing. I also tried a merge and still no conflict. Github is making them up I guess.
> Rebased regardless.

The conflict is due to a rebase and `git` is currently smarter [or less conservative you could say] than the current GitHub merge algorithm.

**Comment by ppedrot** [comment]
Gitlab seems to have failed for a stupid reason, and this was working before. I'll merge this PR anyways.

**Comment by Zimmi48** [comment]
For now it's pretty safe to ignore GitLab CI since it's redundant with other tests.

**Pull request merged**
